### PR TITLE
[Arts 423] [Subfeature] - Integrating SC config server into Practitioner service

### PR DESCRIPTION
**Pull request by darrengsmithson**
## Description
Changes:
renamed config-server to be mts-config-server - this is to ensure it is understood to be the only 'mts' config server.
added search-paths to the config-server properties to ensure the newly created directories in the config-server repo are recognised.
Refactored PractitionerConfiguration to get the service working with the .json file held in the git-repo rather than local resources.
Removed any application.properties from practitioner - only leaving the bootstrap property file.
Added webconfig to the pom for practitioner-service
Added the spring.cloud.config.profile and spring.cloud.config.label  to the service bootstrap property file rather than the config-server application.property file - this is to allow the calling service to make the decision what profile/label should be used.
Removed the sc-config-server package as its a duplicate of mts-config-server
Removed application, log4j and liquibase property files form Trial-config-service as they are now pulled from the config-server repo.
Refactored the TrialConfigController to ensure the property values from the config server are used. 

Also in addition to the changes in this repo, I added service specific directories to the /spring-config-server repo and moved the service specific properties to the relevant directories. The structure of that repo will be default/shared properties held at the base directory with service specific property files held under the service specific directories.

I tested the changes using Postman and all worked ok, I also tested changing the property files held in the /spring-config-server repo and ensured they were picked up without restarting the server. This proves the @RefreshScope is working as expected.

I had to change some JUnit tests due to the refactoring, I havent ensured the tests cover the refactored code so some additional work will be required for that but that is out of the scope of this story. That should be included in the integration of the config-server to existing MTS services story(s).

### CODE REVIEW

File: mts-config-server/pom.xml [hunk 1]
```diff
@@ -1,23 +1,22 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (eladiw) [inline mts-config-server/pom.xml:1]: not sure the new name is easier to understand. IMO sc-xxx means this is part if the spring cloud set of microservices (together with discovery and gateway). wdyt

File: mts-config-server/pom.xml [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
     </parent>
-    <artifactId>sc-config-server</artifactId>
-    <name>sc-config-server</name>
-    <description>Spring Cloud configuration server.
-    </description>
+    <groupId>uk.ac.ox.ndph.mts</groupId>
+    <artifactId>mts-config-server</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <name>mts-config-server</name>
+    <description>MTS Config Server</description>
```
Comment (eladiw) [inline mts-config-server/pom.xml:15]: I prefer to revert this file (maybe except the relative path part)

File: mts-config-server/src/test/java/uk/ac/ox/ndph/mts/configserver/MtsConfigServerApplicationTests.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package uk.ac.ox.ndph.mts.springconfigserver;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class MtsConfigServerApplicationTests {
+
+    @Test
+    void contextLoads() {
```
Comment (eladiw) [inline mts-config-server/src/test/java/uk/ac/ox/ndph/mts/configserver/MtsConfigServerApplicationTests.java:10]: we also need to add actual tests

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java [hunk 4]
```diff
@@ -1,38 +1,58 @@
...
+        } catch (WebClientException wceEx) {
+            throw new RestException(wceEx.getMessage());
         }
-        return configuration;
+
+        return practitionerConfiguration;
+
     }
+
+    private String repoURL() {
```
Comment (eladiw) [inline practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:54]: getConfigRepoUrl()?

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java [hunk 5]
```diff
@@ -1,38 +1,58 @@
...
+
+        try {
+            practitionerConfiguration = WebClient.create().get()
+                .uri(repoURL() + File.separator + "practitioner-configuration.json")
+                .accept(MediaType.APPLICATION_JSON)
+                .retrieve()
+                .bodyToMono(PractitionerConfiguration.class)
+                .block();
+        } catch (WebClientException wceEx) {
+            throw new RestException(wceEx.getMessage());
```
Comment (eladiw) [inline practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:47]: log?

Comment (eladiw) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:47]: wcExc
or wce

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java [hunk 6]
```diff
@@ -1,38 +1,58 @@
...
-import org.springframework.core.io.Resource;
+import org.springframework.http.MediaType;
 import org.springframework.stereotype.Component;
-
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.reactive.function.client.WebClientException;
+import uk.ac.ox.ndph.mts.practitioner_service.exception.RestException;
 import uk.ac.ox.ndph.mts.practitioner_service.model.PractitionerConfiguration;
-import uk.ac.ox.ndph.mts.practitioner_service.exception.InitialisationError;
+import java.io.File;
```
Comment (eladiw) [inline practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:10]: is this needed just for the File.separator?

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java [hunk 7]
```diff
@@ -1,38 +1,58 @@
...
-            try {
-                String jsonString = new String(configurationFile.getInputStream().readAllBytes());
-                configuration = new ObjectMapper().readValue(jsonString, PractitionerConfiguration.class);
-            } catch (Exception e) {
-                throw new InitialisationError(Configurations.ERROR.message(), e);
-            }
+
+        PractitionerConfiguration practitionerConfiguration;
+
+        try {
```
Comment (eladiw) [inline practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:39]: how would that work locally?
will it just use the git as well? or can we have also support for local file for these cases?

File: practitioner-service/src/main/resources/bootstrap.properties [hunk 8]
```diff
@@ -0,0 +1,10 @@
+spring.application.name=practitioner-service
+
+spring.cloud.config.uri=http://localhost:8888
+
+spring.cloud.config.profile=default
+spring.cloud.config.label=main
+
+
```
Comment (eladiw) [inline practitioner-service/src/main/resources/bootstrap.properties:8]: redundant empty lines

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java [hunk 9]
```diff
@@ -12,95 +14,89 @@
...
-    public TrialConfigController(TrialConfigService trialConfigService, String baseUrl) {
-        this.trialConfigService = trialConfigService;
-        this.webClient = WebClient.create(baseUrl);
     }
 
     //TODO(darrensmithson) - we are using a GET and POST for similar functionlaity to accomodate
     // early testing needs, once we are closer to final developmemnt we should ensure only 1 POST is used.
     @GetMapping("/trial")
-    public Trial saveTrialFromGit(
+    public Trial saveTrialFromConfigServer(
```
Comment (eladiw) [inline trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:55]: I prefer all changes to trial config server will be reverted.
a) its going refactoring these days. so we dont want to conflict
b) this method should be deleted. the trial conf service is not exposing endpoint (this will be part of the refactor)

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java [hunk 10]
```diff
@@ -12,95 +14,89 @@
...
-    protected Mono<Trial> createTrialFromURL(String trialConfig) {
-        Mono<Trial> response;
+    protected Trial createTrial(String trialConfig, String url) {
+        Trial trial;
 
         try {
-            response = webClient.get()
-                .uri(trialConfig)
+            trial = webClient.get()
+                .uri(url + File.separator + trialConfig)
```
Comment (eladiw) [inline trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:86]: 1. imo, revert so it won't conflict with the refactoring efforts.
2. same comment about local dev as in the practitioner service

File: trial-config-service/src/main/resources/bootstrap.properties [hunk 11]
```diff
@@ -2,3 +2,9 @@ spring.application.name=trial-config-service
 
 spring.cloud.config.uri=http://localhost:8888
 
+spring.cloud.config.profile=default
+spring.cloud.config.label=main
+
+
```
Comment (eladiw) [inline trial-config-service/src/main/resources/bootstrap.properties:8]: redundant new lines

**Review by eladiw — CHANGES_REQUESTED** [review]
Some comments

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 7 above)
Comment (eladiw) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:39]: can you add this comment:
// Asking the configuration server for the json file

File: mts-config-server/pom.xml [hunk 12]
```diff
@@ -1,23 +1,22 @@
...
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
     <parent>
-        <groupId>uk.ac.ox.ndph.mts</groupId>
-        <artifactId>mts-services-parent</artifactId>
-        <version>0.0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.4.2</version>
```
Comment (eladiw) [inline mts-config-server/pom.xml:8]: we are using 2.3.6 ... this version is not compatible with the azure client
take a look at the 'parent' pom

File: mts-config-server/pom.xml (hunk 1 above)
Comment (darrengsmithson) [reply mts-config-server/pom.xml:1]: The name should be MTS-xx as it is a server that servers MTS config files not spring files. We shouldnt have references to the 3rd party libraries that we use to help build/wire the app....

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 4 above)
Comment (darrengsmithson) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:54]: yep - we can rename this

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 5 above)
Comment (darrengsmithson) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:47]: yep a typo - rename

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 7 above)
Comment (darrengsmithson) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:39]: why do we want it to work 'locally' - i,e with resources - I thought the point of having the config server was to decentralize config files?

File: practitioner-service/src/main/resources/bootstrap.properties (hunk 8 above)
Comment (darrengsmithson) [reply practitioner-service/src/main/resources/bootstrap.properties:8]: not sure why this matters as its the end of the file? It would matter more if it was in a file that had further code as readability would be effected but this is the end of the file?

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 9 above)
Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:55]: I think if we decide to remove any existing endpoints it needs to be a separate unit of work - we should keep this endpoint as it is for this story. The renaming of the method is clarification that we are now not u sing the gitrepo directly but using the config server - changing the name is purely for readability purposes as it will never be called - only the endpoint..

Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:55]: Also just to clarify - the refactoring is necessary to get the controllers working with the config server - when we merge to main these changes will be needed.

File: mts-config-server/pom.xml (hunk 2 above)
Comment (darrengsmithson) [reply mts-config-server/pom.xml:15]: If we keep MTS-xx it will be necessary?

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 6 above)
Comment (darrengsmithson) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:10]: yes, Its needed for the separator

File: trial-config-service/src/main/resources/bootstrap.properties (hunk 11 above)
Comment (darrengsmithson) [reply trial-config-service/src/main/resources/bootstrap.properties:8]: not sure why this matters as its the end of the file? It would matter more if it was in a file that had further code as readability would be effected but this is the end of the file?

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 10 above)
Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:86]: I don't think we should be concerned with refactoring efforts for this merge into the 341 branch - even if we decide to NOT include any code refactoring to existing Controller classes we can make that decision when we decide to merge 341 into Main

Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:86]: I don't think we should be concerned with refactoring efforts for this merge into the 341 branch - even if we decide to NOT include any code refactoring to existing Controller classes we can make that decision when we decide to merge 341 into Main

File: mts-config-server/src/test/java/uk/ac/ox/ndph/mts/configserver/MtsConfigServerApplicationTests.java (hunk 3 above)
Comment (darrengsmithson) [reply mts-config-server/src/test/java/uk/ac/ox/ndph/mts/configserver/MtsConfigServerApplicationTests.java:10]: agreed

File: mts-config-server/pom.xml (hunk 1 above)
Comment (eladiw) [reply mts-config-server/pom.xml:1]: as I see it, its not about what it serves. its about what kind of service this is.
and in this case its not our 'custom application' it is a supportive piece of service. which in this case is a spring cloud type.
everything in this project is MTS. so it doesnt add info. 
you can look at it the same way as the arch diagram, in it, the micro services are grouped to 'custom' apps and sc apps.

File: practitioner-service/src/main/resources/bootstrap.properties (hunk 8 above)
Comment (eladiw) [reply practitioner-service/src/main/resources/bootstrap.properties:8]: doesnt matter. just a minor style comment

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 9 above)
Comment (eladiw) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:55]: yes I agree. my point is that the team is already working on the init service (trial config service). so all these changes should be reverted.

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 10 above)
Comment (eladiw) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:86]: I'm ok with leaving it, if you prefer to.
however I think it will cost in a problematic merge, also I am not sure yet how will the git config repo will look like. and by having 2 clients (trial config and practitioner) we might have to refactor two instead of one microservice.
I would keep this PR to minimum. up to you

File: mts-config-server/pom.xml (hunk 1 above)
Comment (darrengsmithson) [reply mts-config-server/pom.xml:1]: sorry but I still disagree :) - this is an MTS server, sure it uses spring but so do all the classes and we dont name them springcontroller etc - just like you wouldnt name domain classes that rely on Hibernate hibernatePersonxx etc etc. If we had more than 1 project that used a config server which relied on spring would all the config servers be named the same? I dont think they should..they should be named for their purpose which is to server project related files etc

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 9 above)
Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:55]: I think they should go into 341 and then we make a decision to include them in main - keeping the changes in the controllers lets the devs know how to make the config server 'work'. If we decide against merging them into main we should keep the branch for reference, that said, if we are to rework trial-config server why not keep the changes any way as they will be refactored - or possibly completely deleted? I think we need a ref how we retrieve the files..it will save some time for somebody.

File: trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java (hunk 10 above)
Comment (darrengsmithson) [reply trial-config-service/src/main/java/uk/ac/ox/ndph/mts/trial_config_service/controller/TrialConfigController.java:86]: still think refactoring should remain. I think Practitioner will be refactored soon anyway

File: practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java (hunk 7 above)
Comment (eladiw) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:39]: you are right its not the place I meant to comment in.
what I meant is that the config server will , in local mode, wont go to git, instead it will use a file...

Comment (darrengsmithson) [reply practitioner-service/src/main/java/uk/ac/ox/ndph/mts/practitioner_service/configuration/PractitionerConfigurationProvider.java:39]: why (in what scenarios) would we use 'local' resources when using the config server?

File: mts-config-server/pom.xml (hunk 1 above)
Comment (eladiw) [reply mts-config-server/pom.xml:1]: talked offline. we'll remove the prefix

File: config-server/pom.xml [hunk 13]
```diff
@@ -1,23 +1,22 @@
...
-    <description>Spring Cloud configuration server.
-    </description>
+    <groupId>uk.ac.ox.ndph.mts</groupId>
+    <artifactId>config-server</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <name>config-server</name>
+    <description>Spring Cloud Configuration Server</description>
     <properties>
-        <azure.version>2.3.5</azure.version>
+        <java.version>15</java.version>
```
Comment (eladiw) [inline config-server/pom.xml:17]: java ver 15 is not correct. need to revert this change

**Review by eladiw — APPROVED** [review]

**Review by anatbal — APPROVED** [review]

**Comment by darrengsmithson** [comment]
merging after Elad/Anats reviews
